server/services/v1/auth.go

```diff
@@ -42,53 +52,194 @@ type authService struct {
 }
 
 type OAuthProvider interface {
-	GetAccessToken(request *api.GetAccessTokenRequest) (*api.GetAccessTokenResponse, error)
+	GetAccessToken(req *api.GetAccessTokenRequest) (*api.GetAccessTokenResponse, error)
+	CreateApplication(ctx context.Context, req *api.CreateApplicationRequest) (*api.CreateApplicationResponse, error)
+	RotateApplicationSecret(ctx context.Context, req *api.RotateApplicationSecretRequest) (*api.RotateApplicationSecretResponse, error)
+	DeleteApplication(ctx context.Context, req *api.DeleteApplicationsRequest) (*api.DeleteApplicationResponse, error)
+	ListApplications(ctx context.Context, req *api.ListApplicationsRequest) (*api.ListApplicationsResponse, error)
 }
 
 type Auth0 struct {
-	config.AuthConfig
+	authConfig config.AuthConfig
+	management *management.Management
 }
 
-func (a Auth0) GetAccessToken(req *api.GetAccessTokenRequest) (*api.GetAccessTokenResponse, error) {
-	data := url.Values{
-		"refresh_token": {req.RefreshToken},
-		"client_id":     {a.AuthConfig.ClientId},
-		"grant_type":    {refreshToken},
-		"scope":         {scope},
+func (a *Auth0) GetAccessToken(req *api.GetAccessTokenRequest) (*api.GetAccessTokenResponse, error) {
+	switch req.GrantType {
+	case api.GrantType_REFRESH_TOKEN:
+		return getAccessTokenUsingRefreshToken(req, a)
+	case api.GrantType_CLIENT_CREDENTIALS:
+		return getAccessTokenUsingClientCredentials(req, a)
 	}
-	resp, err := http.PostForm(a.AuthConfig.ExternalTokenURL, data)
+	return nil, api.Errorf(api.Code_INVALID_ARGUMENT, "Failed to GetAccessToken: reason = unsupported grant_type, it has to be one of [refresh_token, client_credentials]")
+}
+
+func (a *Auth0) CreateApplication(ctx context.Context, req *api.CreateApplicationRequest) (*api.CreateApplicationResponse, error) {
+	currentSub, err := getCurrentSub(ctx)
 	if err != nil {
-		return nil, api.Errorf(api.Code_INTERNAL, "Failed to get access token: reason = "+err.Error())
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to list applications: reason = %s", err.Error())
 	}
-	defer resp.Body.Close()
-	body, err := io.ReadAll(resp.Body)
+	currentNamespace, err := request.GetNamespace(ctx)
 	if err != nil {
-		return nil, api.Errorf(api.Code_INTERNAL, "Failed to get access token: reason = "+err.Error())
+		return nil, api.Errorf(api.Code_INTERNAL, "Failed to list applications: reason = %s", err.Error())
 	}
-	bodyStr := string(body)
-	if resp.StatusCode == 200 {
-		getAccessTokenResponse := api.GetAccessTokenResponse{}
 
-		err = json.Unmarshal([]byte(bodyStr), &getAccessTokenResponse)
+	nonInteractiveApp := "non_interactive"
+	metadata := map[string]string{}
+	metadata[createdAt] = strconv.FormatInt(time.Now().UnixMilli(), 10)
+	metadata[createdBy] = currentSub
+	metadata[tigrisNamespace] = currentNamespace
+
+	c := &management.Client{
+		Name:           &req.Name,
+		Description:    &req.Description,
+		AppType:        &nonInteractiveApp,
+		ClientMetadata: metadata,
+	}
+
+	err = a.management.Client.Create(c)
```

Comment:
What happens if it is already created? i.e. leftover from the last time. The case when the requests didn't reach the cleanup flow. Do they return some error that you can use to identify that case and then we don't need to return from here but try to grant again.

Reply:
there is no way to detect if this is the same application being created again because id is auto-generated. We can improve this later and limit the unique name per account/org. wdyt?

Reply:
Are there any other metadata attached to it? We can use that to read first.